rq/queue.py

```diff
@@ -252,33 +252,35 @@ def enqueue_call(self, func, args=None, kwargs=None, timeout=None,
             depends_on=depends_on, timeout=timeout, id=job_id,
             origin=self.name, meta=meta)
 
-        # If job depends on an unfinished job, register itself on it's
-        # parent's dependents instead of enqueueing it.
-        # If WatchError is raised in the process, that means something else is
-        # modifying the dependency. In this case we simply retry
+        # If a _dependent_ job depends on any unfinished job, register all the
+        #_dependent_ job's dependencies instead of enqueueing it.
+        #
+        # `Job#fetch_dependencies` sets WATCH on all dependencies. If
+        # WatchError is raised in the when the pipeline is executed, that means
+        # something else has modified either the set of dependencies or the
+        # status of one of them. In this case, we simply retry.
         if depends_on is not None:
-            if not isinstance(depends_on, self.job_class):
-                depends_on = self.job_class(id=depends_on,
-                                            connection=self.connection)
             with self.connection.pipeline() as pipe:
                 while True:
                     try:
-                        pipe.watch(depends_on.key)
-
-                        # If the dependency does not exist, raise an
-                        # exception to avoid creating an orphaned job.
-                        if not self.job_class.exists(depends_on.id,
-                                                     self.connection):
-                            raise InvalidJobDependency('Job {0} does not exist'.format(depends_on.id))
-
-                        if depends_on.get_status() != JobStatus.FINISHED:
-                            pipe.multi()
-                            job.set_status(JobStatus.DEFERRED)
-                            job.register_dependency(pipeline=pipe)
-                            job.save(pipeline=pipe)
-                            job.cleanup(ttl=job.ttl, pipeline=pipe)
-                            pipe.execute()
-                            return job
+                        dependencies = job.fetch_dependencies(
+                            watch=True,
+                            pipeline=pipe
+                        )
+
+                        pipe.watch(job.dependencies_key)
+
+                        pipe.multi()
+
+                        for dependency in dependencies:
+                            if dependency.get_status() != JobStatus.FINISHED:
+                                job.set_status(JobStatus.DEFERRED)
+                                job.register_dependency(pipeline=pipe)
+                                job.save(pipeline=pipe)
+                                job.cleanup(ttl=job.ttl, pipeline=pipe)
+                                pipe.execute()
```

Comment:
I think we can call `pipe.execute()` outside of the for loop to write everything all at once.

Reply:
@selwin do you mind taking another glance? The way I read lines 275 - 282 is:

- Iterate over all of `job`'s dependencies
- IF any _one_ dependency is not `FINISHED`, THEN: register dependencies in redis, set the job to `DEFERRED`, `save`, `cleanup` and `return`(_short circuiting_ `enqueue_call`).

Moving it outside side of the loop seems like it would necessitate changing the _short circuit_ structure the method.

Reply:
Yes, you're absolutely right. That was my mistake.